Implementation for feature #15

sspat opened the pull request:
Please review my pull request.
I am sorry if i made any mistakes, this is my first attempt at open source contribution.

sspat committed:
Implementation for feature #15

sspat committed:
Fixed bug when trying to save junction table rows for related models that were not saved before.

sspat committed:
Updated tests for feature #15 implementation coverage.

sspat committed:
Fixed assertion description in test for feature #15

juban commented:
Hi. Thank you for your contribution to this project. I will try to review your code as soon as possible. I'm very busy right now so it may take a few days. Thanks again.

sspat committed:
Added a new mode to the _computePkDiff method to allow forcing removal of all rows in a junction table representing the initial state of a many-to-many relation.
This mode is automatically used if junction table properties are configured for the relation, ensuring that junction table properties are saved even if the related model properties did not change.
feature #15

sspat committed:
Fixed bug for feature #15 where related records could not be found on $actualModels generation.

juban commented:
Hi. I reviewed your implementation and it's a very nice feature addition, thanks. You did a great job.

Anyway, I have a few questions/suggestions:

- I'm not too excited about adding a new public property. I would rather prefer that we keep the single relation array property and add the junctionTableColumns as an associative property per relation. Based on your exemple, it would look like that:
`'relations' => [
                    'company',
                    'users',
                    'links',
                    'tags' => [
                        'junctionTableColumns' => function ($model) {
                            return [
                                'order' => $model->getOrder()
                            ];
                        }
                    ]
                ],` 
That would be great in case we need to add more per relation parameters in the future.
- _computePkDiff refactoring lacks the ability to only re-save changed related models and additional fields instead of deleting and save them all again. I guess this could be achieved by comparing old an new additional columns.

What do you think?

sspat commented:
Hi! Thank you very much for your time and evaluation!

I agree on the configuration, i will make an implementation right away, taking care that it does not brake backwards compatibility.

On behalf of _computePkDiff - maybe you can give an advice, the problem is that i could not find any Yii2 out-of-box way to fetch those junction table properties and store them in the ActiveRecord instance to allow access to their values and compare them to the new values. I myself am using this extension for working with them - https://github.com/alexinator1/yii2-jta, but i think it's not a good idea to make this behavior depend on that extension. Maybe there is some other more simple solution i am not taking into account?

juban commented:
Regarding the associative configuration, that's kind of part of my roadmap too. So if you can help on that, thanks, that will speed things a bit :)

For the junction column part, I agree it will be tricky. I'm definitely not a fan of being dependent on an other extension. I will try to figure out a way to do that tomorrow and let you know if I've found anything of interest. If it seems too complicated, we could go with your implementation as a first step. 

Thanks again for your time and support.

juban commented:
Hi @sspat,

I just pushed some changes to the master branch to be able to set per relation properties as discussed.
This will be helpful I hope to implement the junctionTableColumns property. Could you merge your code and use that? This could be done with ease I think, but please let me know if you find any issue.

Regarding the junction properties values comparison thing, I did not find the time to investigate yet. Did you made any progress on your side ?

Thanks.

sspat commented:
Hi @juban !

I am sorry I could not give the attention i promised to the project, got caught up in my work.
I will allocate some time in the next couple of days to make progress on this feature.
Thank you for not closing it and keeping up your great work on this project, i am using it very extensively!

juban commented:
Hi @sspat 

I've just pushed an implementation of your feature into the `extra-column-junction-table` branch.
I've made some minor adjustments as discussed for the configuration part, but that's mainly based on your excellent work.
Could you review that and let me know if it seems fine to you?

Thanks.

sspat commented:
Hi @juban !

I reviewed the branch you mentioned and it completely solves my use-case and you took a much cleaner and universal approach to the configuration part.
I am happy that my contribution was useful.
If there is anything more i can do on this case please let me know.

juban commented:
Hi @sspat ,

Thank you for your feedback. I'm happy it suites your use case.
You did most of the work for that feature anyway, so thank you again for your contribution.
I will release a 1.4.0 release shortly that will include the new feature.

1allen commented on src/SaveRelationsBehavior.php:
```diff
@@ -298,29 +299,40 @@ public function afterSave()
                         });
+                        $initialRelations = $model->{$relationName};
                         foreach ($deletedPks as $key) {
                             $model->unlink($relationName, $initialModels[$key], true);
```
btw this `unlink()` is a bit redundant. it does not take `$junctionTableColumns` into account, so it can delete all the relations to a single entry.
see https://github.com/yiisoft/yii2/issues/14598

juban replied:
@1allen What do you mean by "redundant"?

1allen replied:
@juban I mean that it could unlink more records than actually expected. You can see an example in the yii2 issue I've mentioned above. Sorry, I've no idea how to fix it at this moment

juban replied:
@1allen Indeed, the behavior cannot handle that kind of complex scenario. And as long as the unlink method doesn't handle it either, I don't see an easy way to solve this without jeopardizing the behavior compatibility for different kind of DBMS.
